### PR TITLE
feat: add json_fields extras argument for adding to jsonPayload

### PR DESCRIPTION
**Pull request by daniel-sanche**
v3.0.0 will support sending jsonPayload log entries by logging a dict object: `cloud_logger.error(my_dict)`. This feature is technically counter to the python logging spec though, which expects string inputs. For full compatibility, we advise users to send their dicts encoded as JSON like this: `cloud_logger.error(json.dumps(my_dict))`

These methods of sending JSON payloads [don't work for all customers though](https://github.com/googleapis/python-logging/issues/411), so this PR adds a third method: users can now send json data through the `extras` argument: `cloud_logger.error('bad news', extra={json_fields: my_dict}))`. Logging this way will result in a jsonPayload log, made up of the `json_fields` dict combined with the log message.

Fixes https://github.com/googleapis/python-logging/issues/411

### CODE REVIEW

File: google/cloud/logging_v2/client.py [hunk 1]
```diff
@@ -376,7 +376,7 @@ def get_default_handler(self, **kw):
             if monitored_resource.type == _GAE_RESOURCE_TYPE:
                 return CloudLoggingHandler(self, resource=monitored_resource, **kw)
             elif monitored_resource.type == _GKE_RESOURCE_TYPE:
-                return ContainerEngineHandler(**kw)
+                return StructuredLogHandler(**kw, project_id=self.project)
```
Comment (minherz) [inline google/cloud/logging_v2/client.py:379]: nit: I think that I saw this commit in PR #450. Is it possible there had to be a rebase?

File: google/cloud/logging_v2/handlers/handlers.py [hunk 2]
```diff
@@ -221,9 +221,16 @@ def _format_and_parse_message(record, formatter_handler):
         record (logging.LogRecord): The record object representing the log
         formatter_handler (logging.Handler): The handler used to format the log
     """
+    passed_json_fields = getattr(record, "json_fields", {})
     # if message is a dictionary, return as-is
     if isinstance(record.msg, collections.abc.Mapping):
-        return record.msg
+        payload = record.msg
+        if passed_json_fields and isinstance(
+            passed_json_fields, collections.abc.Mapping
```
Comment (minherz) [inline google/cloud/logging_v2/handlers/handlers.py:229]: nit: if passed_json_fields is undefined (i.e. `if not passed_json_fields` then the line 226 probably will fail. should we check for it sooner?

File: tests/unit/handlers/test_structured_log.py [hunk 3]
```diff
@@ -321,3 +321,26 @@ def test_format_overrides(self):
...
+    def test_format_with_json_fields(self):
+        """
+        User can add json_fields to the record, which should populate the payload
+        """
+        import logging
+        import json
+
+        handler = self._make_one()
+        message = "name: %s"
+        name_arg = "Daniel"
```
Comment (minherz) [inline tests/unit/handlers/test_structured_log.py:334]: question: i am not familiar with the interface, is it supposed to format the message like with print() ?

**Review by minherz — APPROVED** [review]
lgtm. i did not do in-depth analysis of system and unit tests. maybe someone will look into it latter..

**Review by losalex — APPROVED** [review]
LGTM

File: google/cloud/logging_v2/client.py (hunk 1 above)
Comment (daniel-sanche) [reply google/cloud/logging_v2/client.py:379]: Yeah, it seems like there was a merging issue in the v3.0.0 branch at some point which changed this line, and I had to fix it in both branches for tests to pass

File: google/cloud/logging_v2/handlers/handlers.py (hunk 2 above)
Comment (daniel-sanche) [reply google/cloud/logging_v2/handlers/handlers.py:229]: I'm not sure I understand. Line 226 should be unrelated to `passed_json_fields`. It's checking if the `msg` field was populated by a dictionary. If so, lines 228-232 will attempt to add the additional `passed_json_fields` fields to the msg dictionary (if present)

File: tests/unit/handlers/test_structured_log.py (hunk 3 above)
Comment (daniel-sanche) [reply tests/unit/handlers/test_structured_log.py:334]: The idea is that Python's logging std lib [allows you to build formated log strings](https://docs.python.org/3/howto/logging-cookbook.html#use-of-alternative-formatting-styles) like this: `logging.info("my name is: %s", "Daniel")`. This test just makes sure that our GCP library additions don't break that expected functionality

File: google/cloud/logging_v2/handlers/handlers.py (hunk 2 above)
Comment (daniel-sanche) [reply google/cloud/logging_v2/handlers/handlers.py:230]: I tried to improve the comments here

**Pull request merged**